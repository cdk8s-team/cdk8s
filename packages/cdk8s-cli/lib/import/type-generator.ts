import { JSONSchema4 } from "json-schema";
import { CodeMaker, toPascalCase } from "codemaker";

const PRIMITIVE_TYPES = [ 'string', 'number', 'integer', 'boolean' ];
const DEFINITIONS_PREFIX = '#/definitions/';

export interface TypeGeneratorOptions {
  exclude?: string[];
}

export interface GeneratedConstruct {
  readonly fqn: string;
  readonly group: string;
  readonly version: string;
  readonly kind: string;
  readonly schema: JSONSchema4;
}

export class TypeGenerator {
  private readonly typesToEmit: { [name: string]: (code: CodeMaker) => void } = { };
  private readonly emittedTypes = new Set<string>();
  private readonly exclude: string[];

  constructor(private readonly schema: JSONSchema4 = { }, options: TypeGeneratorOptions = { }) {
    this.exclude = options.exclude || [];
  }

  public emitConstruct(def: GeneratedConstruct) {
    this.emitLater(def.kind, code => {
      const options = createOptionsStructSchema();

      const optionsStructName = options.type ? `${def.kind}Options` : 'any';
      const schema = def.schema;

      this.emitType(optionsStructName, options, def.fqn);

      emitConstruct();

      function createOptionsStructSchema() {
        const copy: JSONSchema4 = { ...def.schema || {} };
        const props = copy.properties = copy.properties || {};
        delete props.apiVersion;
        delete props.kind;
        delete props.status;
        delete copy['x-kubernetes-group-version-kind'];
    
        copy.required = copy.required || [];
        copy.required = copy.required.filter(x => x !== 'apiVersion' && x !== 'kind' && x !== 'status');
    
        return copy;
      }
    
      function emitConstruct() {
        code.line('/**');
        code.line(` * ${def.schema?.description ?? ''}`);
        code.line(` *`);
        code.line(` * @schema ${def.fqn}`)
        code.line(` */`);
        code.openBlock(`export class ${def.kind} extends ApiObject`);
    
        emitInitializer();
      
        code.closeBlock();
      }
    
      function emitInitializer() {
  
        code.line(`/**`);
        code.line(` * Defines a "${def.fqn}" API object`);
        code.line(` * @param scope the scope in which to define this object`);
        code.line(` * @param name a scope-local name for the object`);
        code.line(` * @param options configuration options`);
        code.line(` */`);
    
        const hasRequired = schema?.required && Array.isArray(schema.required) && schema.required.length > 0;
        const defaultOptions = hasRequired ? '' : ' = {}';
        code.openBlock(`public constructor(scope: Construct, name: string, options: ${optionsStructName}${defaultOptions})`);
        emitInitializerSuper();
    
        code.closeBlock();
      }
    
      function emitInitializerSuper() {
        const groupPrefix = def.group ? `${def.group}/` : '';
        code.open(`super(scope, name, {`);
        code.line(`...options,`);
        code.line(`kind: '${def.kind}',`);
        code.line(`apiVersion: '${groupPrefix}${def.version}',`);
        code.close(`});`);    
      }
    });
  }

  public emitType(typeName: string, def: JSONSchema4, structFqn: string): string {
    typeName = normalizeTypeName(typeName);

    if (structFqn.startsWith(DEFINITIONS_PREFIX)) {
      structFqn = structFqn.substring(DEFINITIONS_PREFIX.length);
    }

    if (this.isExcluded(structFqn)) {
      throw new Error(`Type ${structFqn} cannot be added since it matches one of the exclusion patterns`);
    }

    // skip api objects (they are emitted as constructs and not as data types)
    if ('x-kubernetes-group-version-kind' in def && def.properties?.metadata) {
      return typeName;
    }

    if (def.$ref) {
      return this.typeForRef(def);
    }

    // unions
    if (def.oneOf || def.anyOf) {
      this.emitUnion(typeName, def, structFqn)
      return typeName;
    }

    if (def.type === 'string' && def.format === 'date-time') {
      return `Date`;
    }
  
    switch (def.type) {
      case undefined: return 'string';
      case 'boolean': return 'boolean';
      case 'array': return `${this.typeForArray(typeName, def)}[]`;
      case 'any': return 'any';
      case 'null': return 'any';
    }

    if (def.type === 'number' || def.type === 'integer') {
      return 'number';
    }

    if (def.type === 'string') {
      if (def.format === 'date-time') {
        return 'Date';
      }

      return 'string';
    }
    
    if (def.type !== 'object') {
      throw new Error(`unexpected schema type ${def.type}. Expecting "object"`);
    }

    // map
    if (!def.properties && def.additionalProperties && typeof(def.additionalProperties) === 'object') {
      return `{ [key: string]: ${this.typeForProperty(typeName, def.additionalProperties)} }`;
    }

    // struct
    if (def.properties) {
      this.emitStruct(typeName, def, structFqn)
      return typeName;
    }

    return 'any';
  }

  public generate(code: CodeMaker) {
    while (Object.keys(this.typesToEmit).length) {
      const name = Object.keys(this.typesToEmit)[0];
      const emitter = this.typesToEmit[name];
      emitter(code);
      code.line();
      delete this.typesToEmit[name];
      this.emittedTypes.add(name);
    }    
  }

  private emitLater(typeName: string, codeEmitter: (code: CodeMaker) => void) {
    if (this.emittedTypes.has(typeName)) {
      return;
    }

    this.typesToEmit[typeName] = codeEmitter;
  }

  private emitUnion(typeName: string, def: JSONSchema4, fqn: string) {
    this.emitLater(typeName, code => {
      this.emitDescription(code, fqn, def.description);

      code.openBlock(`export class ${typeName}`);

      for (const option of def.oneOf || def.anyOf || []) {
        if (typeof(option.type) !== 'string' || !PRIMITIVE_TYPES.includes(option.type)) {
          throw new Error(`unexpected union type ${JSON.stringify(option.type)}`);
        }

        const type = option.type === 'integer' ? 'number' : option.type;
        const methodName = 'from' + type[0].toUpperCase() + type.substr(1);
        code.openBlock(`public static ${methodName}(value: ${type}): ${typeName}`);
        code.line(`return new ${typeName}(value);`);
        code.closeBlock();
      }

      code.openBlock(`private constructor(value: any)`);
      code.line(`Object.defineProperty(this, 'resolve', { value: () => value });`);
      code.closeBlock();

      code.closeBlock();
    });
  }

  private emitStruct(typeName: string, structDef: JSONSchema4, structFqn: string) {
    this.emitLater(typeName, code => {
      this.emitDescription(code, structFqn, structDef.description);
      code.openBlock(`export interface ${typeName}`);

      for (const [ propName, propSpec ] of Object.entries(structDef.properties || {})) {
  
        if (propName.startsWith('x-')) {
          continue; // skip extensions for now
        }
  
        this.emitProperty(code, propName, propSpec, structFqn, structDef);
      }
    
      code.closeBlock();
    });
  }

  private emitProperty(code: CodeMaker, name: string, propDef: JSONSchema4, structFqn: string, structDef: JSONSchema4) {
    const originalName = name;

    // if name is not camelCase, convert it to camel case, but this is likely to
    // produce invalid output during synthesis, so add some annotation to the docs.
    if (name[0] === name[0].toUpperCase()) {
      name = code.toCamelCase(name);
    }

    // if the name starts with '$' (like $ref or $schema), we remove the "$"
    // and it's the same deal - will produce invalid output
    if (name.startsWith('$')) {
      name = name.substring(1);
    }

    this.emitDescription(code, `${structFqn}#${originalName}`, propDef.description);
    const propertyType = this.typeForProperty(`${structFqn}.${name}`, propDef);
    const required = this.isPropertyRequired(name, structDef);
    const optional = required ? '' : '?';

    code.line(`readonly ${name}${optional}: ${propertyType};`);
    code.line();
  }

  private emitDescription(code: CodeMaker, fqn: string, description?: string, annotations: { [type: string]: string } = { }) {
    code.line('/**');

    if (description) {
      description = description.replace(/\*\//g, '_/');

      const extractDefault = /Defaults?\W+(to|is)\W+(.+)/g.exec(description);
      const def = extractDefault && extractDefault[2];
    
      code.line(` * ${description}`);
      if (def) {
        annotations['default'] = def;
      }

      code.line(' *');
    }

    annotations['schema'] = fqn;

    for (const [ type, value ] of Object.entries(annotations)) {
      code.line(` * @${type} ${value}`);
    }

    code.line(' */')
  }

  private typeForProperty(propertyFqn: string, def: JSONSchema4): string {
    const subtype = propertyFqn.split('.').map(x => toPascalCase(x)).join('');
    return this.emitType(subtype, def, subtype);
  }

  private typeForRef(def: JSONSchema4): string {
    const prefix = '#/definitions/';
    if (!def.$ref || !def.$ref.startsWith(prefix)) {
      throw new Error(`invalid $ref`);
    }

    if (this.isExcluded(def.$ref)) {
      return 'any';
    }

    const comps = def.$ref.substring(prefix.length).split('.');
    const typeName = comps[comps.length - 1];
    const schema = this.resolveReference(def);
    return this.emitType(typeName, schema, def.$ref);
  }

  private typeForArray(propertyFqn: string, def: JSONSchema4): string {
    if (!def.items || typeof(def.items) !== 'object') {
      throw new Error(`unsupported array type ${def.items}`);
    }

    return this.typeForProperty(propertyFqn, def.items);
  }  

  private resolveReference(def: JSONSchema4): JSONSchema4 {
    const ref = def.$ref;
    if (!ref || !ref.startsWith(DEFINITIONS_PREFIX)) {
      throw new Error(`expecting a local reference`);
    }

    if (!this.schema.definitions) {
      throw new Error(`schema does not have "definitions"`);
    }

    const lookup = ref.substr(DEFINITIONS_PREFIX.length);
    const found = this.schema.definitions[lookup];
    if (!found) {
      throw new Error(`cannot resolve local reference ${ref}`);
    }

    return found;
  }

  private isPropertyRequired(property: string, structDef: JSONSchema4) {
    return Array.isArray(structDef.required) && structDef.required.includes(property);
  }

  private isExcluded(fqn: string) {
    for (const pattern of this.exclude) {
      const re = new RegExp(pattern);
      if (re.test(fqn)) {
        return true;
      }
    }

    return false;
  }
}



function normalizeTypeName(typeName: string) {
  if (!typeName.startsWith('I')) { // avoid the regex
    return typeName;
  }

  // if the type name starts with IXXXFoo then make XXX lowercase because we assume
  // it's an acromym that starts with an I, and this will be identified as an interface by JSII.
  const re = /^I([A-Z]+)([A-Z][a-z]+[A-Za-z]+)$/.exec(typeName);
  if (!re) {
    return typeName;
  }

  const group1 = re[1];
  const rest = re[2];
  const normalized = `I${group1.toLocaleLowerCase()}${rest}`;
  return normalized;
}