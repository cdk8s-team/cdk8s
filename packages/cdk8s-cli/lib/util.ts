import * as http from 'http';
import * as https from 'https';
import { spawn, SpawnOptions } from 'child_process';
import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';

export async function shell(program: string, args: string[] = [], options: SpawnOptions = { }) {
  return new Promise((ok, ko) => {
    const child = spawn(program, args, { stdio: 'inherit', ...options });
    child.once('error', ko);
    child.once('exit', code => {
      if (code === 0) { return ok(); }
      else { return ko(new Error(`non-zero exit code ${code}`)); }
    });
  });
}

export async function withTempDir(dirname: string, closure: () => Promise<void>) {
  const prevdir = process.cwd();
  const parent = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk8s.'));
  const workdir = path.join(parent, dirname);
  await fs.mkdirp(workdir);
  try {
    process.chdir(workdir);
    await closure();
  } finally {
    process.chdir(prevdir);
    await fs.remove(parent);
  }
}

async function get(url: string, protocol: typeof http | typeof https = https): Promise<string> {
  return new Promise((ok, ko) => {
    const req = protocol.get(url, async res => {
      switch(res.statusCode) {
        case 200: {
          const data = new Array<Buffer>();
          res.on('data', chunk => data.push(chunk));
          res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
          res.once('error', ko);
          break;
        }

        case 301:
        case 302: {
          if (res.headers.location) {
            await ok(get(res.headers.location, protocol));
          }
          break;
        }

        default: {
          throw new Error(`${res.statusMessage}: ${url}`);
        }
      }
    });

    req.once('error', ko);
    req.end();
  });
}

export async function httpGet(url: string): Promise<string> {
  return get(url, http)
}

export async function httpsGet(url: string): Promise<string> {
  return get(url)
}
