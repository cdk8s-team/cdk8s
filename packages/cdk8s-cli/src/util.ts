import * as http from 'http';
import * as https from 'https';
import { spawn, SpawnOptions } from 'child_process';
import { parse } from 'url';
import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';

export async function shell(program: string, args: string[] = [], options: SpawnOptions = { }) {
  const command = `"${program} ${args.join(' ')}" at ${path.resolve(options.cwd ?? '.')}`;
  return new Promise((ok, ko) => {
    const child = spawn(program, args, { stdio: 'inherit', ...options });
    child.once('error', err => {
      throw new Error(`command ${command} failed: ${err}`);
    });
    child.once('exit', code => {
      if (code === 0) {
        return ok();
      }
      else {
        return ko(new Error(`command ${command} returned a non-zero exit code ${code}`));
      }
    });
  });
}

export async function mkdtemp(closure: (dir: string) => Promise<void>) {
  const workdir = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk8s-'));
  try {
    await closure(workdir);
  } finally {
    await fs.remove(workdir);
  }
}

export async function download(url: string): Promise<string> {
  let client: typeof http | typeof https;
  const proto = parse(url).protocol;

  if (!proto || proto === 'file:') {
    return await fs.readFile(url, 'utf-8');
  }

  switch (proto) {
    case 'https:':
      client = https;
      break;

    case 'http:':
      client = http;
      break;

    default:
      throw new Error(`unsupported protocol ${proto}`);
  }

  return new Promise((ok, ko) => {
    const req = client.get(url, res => {
      switch(res.statusCode) {
        case 200: {
          const data = new Array<Buffer>();
          res.on('data', chunk => data.push(chunk));
          res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
          res.once('error', ko);
          break;
        }

        case 301:
        case 302: {
          if (res.headers.location) {
            ok(download(res.headers.location));
          }
          break;
        }

        default: {
          throw new Error(`${res.statusMessage}: ${url}`);
        }
      }
    });

    req.once('error', ko);
    req.end();
  });
}
