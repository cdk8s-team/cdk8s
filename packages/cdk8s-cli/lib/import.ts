import * as https from 'https';
import * as fs from 'fs-extra';
import * as path from 'path';
import { promises } from 'fs';
import { emitConstructForApiObject, findApiObjectDefinitions, selectApiObjects } from '../lib/codegen-constructs';
import { CodeMaker } from 'codemaker';
import { JSONSchema4 } from 'json-schema';
import { TypeGenerator } from './codegen-types';
import { withTempDir, shell } from './util';
import { jsiiCompile } from './jsii';

export const DEFAULT_API_VERSION = '1.14.0';

export enum Language {
  TYPESCRIPT = 'typescript',
  PYTHON = 'python',
  DOTNET = 'dotnet',
  JAVA = 'java',
}

export const LANGUAGES = [ Language.TYPESCRIPT, Language.PYTHON ];

export interface Options {
  /**
   * Output programming language.
   */
  readonly language: Language;

  /**
   * The API version to generate.
   */
  readonly apiVersion?: string;

  /**
   * FQNs of API object types to select instead of selecting the latest stable
   * version.
   * 
   * @default - selects the latest stable version from each API object
   */
  readonly include?: string[];

  /**
   * Do not import these types. Instead, represent them as "any".
   * 
   * @default - include all types that derive from the root types.
   */
  readonly exclude?: string[];
}

export async function generateAllApiObjects(outdir: string, options: Options) {
  outdir = path.resolve(outdir);

  if (options.language === Language.TYPESCRIPT) {
    await generateApiObjectsTypeScript(outdir, options);
    console.log(`${outdir}/k8s.ts`);
    return;
  }

  // this is not typescript, so we generate in a staging directory and harvest the code
  await withTempDir('k8s', async () => {
    await generateApiObjectsTypeScript('.', options);
    await jsiiCompile('.');

    const pacmak = require.resolve('jsii-pacmak/bin/jsii-pacmak');
    await shell(pacmak, [ '--target', options.language, '--code-only' ]);
    await harvestCode(options.language, outdir);
  });
}

async function harvestCode(language: Language, outdir: string) {
  switch (language) {
    case Language.TYPESCRIPT:
      throw new Error('no op for typescript');

    case Language.PYTHON:
      await harvestPython();
      break;

    default:
      throw new Error(`unsupported language ${language} (yet)`);
  }

  async function harvestPython() {
    const target = path.join(outdir, 'k8s');
    await fs.move('dist/python/src/k8s', target, { overwrite: true });
    console.error(target);
  }
}

async function generateApiObjectsTypeScript(outdir: string, options: Options) {
  const code = new CodeMaker();
  code.indentation = 2;

  const schema = await downloadSchema(options.apiVersion ?? DEFAULT_API_VERSION);
  const map = findApiObjectDefinitions(schema);

  const topLevelObjects = selectApiObjects(map, { include: options.include });

  code.openFile('k8s.ts');
  code.line(`// generated by cdk8s`);
  code.line(`import { ApiObject } from 'cdk8s';`);
  code.line(`import { Construct } from '@aws-cdk/core';`);
  code.line();

  const typeGenerator = new TypeGenerator(schema, { exclude: options.exclude });

  for (const o of topLevelObjects) {
    emitConstructForApiObject(code, typeGenerator, o);
  }

  typeGenerator.generate(code);

  code.closeFile('k8s.ts');

  await promises.mkdir(outdir, { recursive: true });
  await code.save(outdir);  
}

async function downloadSchema(apiVersion: string) {
  const output = await httpsGet(`https://kubernetesjsonschema.dev/v${apiVersion}/_definitions.json`);
  return JSON.parse(output) as JSONSchema4;
}

async function httpsGet(url: string): Promise<string> {
  return new Promise((ok, ko) => {
    const req = https.get(url, res => {
      if (res.statusCode !== 200) {
        throw new Error(`${res.statusMessage}: ${url}`);
      }
      const data = new Array<Buffer>();
      res.on('data', chunk => data.push(chunk));
      res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
      res.once('error', ko);
    });

    req.once('error', ko);
    req.end();
  });
}
