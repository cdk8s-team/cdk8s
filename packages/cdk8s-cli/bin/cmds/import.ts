import * as yargs from 'yargs';
import { generateAllApiObjects, DEFAULT_API_VERSION } from '../../lib/import';

const K8S_SPEC = 'k8s';
const DEFAULT_OUTDIR = 'imports';

class Command implements yargs.CommandModule {
  public readonly command = 'import SPEC';
  public readonly describe = 'Imports API objects to your app by generating constructs';
  public readonly aliases = [ 'gen', 'import', 'generate' ];

  public readonly builder = (args: yargs.Argv) => args
    .positional('SPEC', {  desc: `The API spec to import (only "k8s" is currently supported)` })
    .demand('SPEC')
    .example(`cdk8s import k8s`, `imports Kubenretes API objects to imports/k8s.ts`)
    .example(`cdk8s import k8s@1.13.0`, `imports a specific version of the Kubenretes API`)

    .option('output', { type: 'string', desc: 'output directory', default: DEFAULT_OUTDIR, alias: 'o' })
    .option('include', { type: 'array', desc: 'types to select instead of the default which is latest stable version' })
    .option('exclude', { type: 'array', desc: 'do not import types that match these regular expressions. They will be represented as the "any" type' });

  public async handler(argv: any) {
    const spec = argv.spec;
    const [ src, specver ] = spec.split('@');

    const apiVersion = specver ?? DEFAULT_API_VERSION;

    if (src !== K8S_SPEC) {
      throw new Error(`currently we only support the source ${K8S_SPEC}`);
    }

    await generateAllApiObjects(argv.output, {
      apiVersion,
      include: argv.include,
      exclude: argv.exclude
    });

    console.error(`k8s@${apiVersion} imported to ${argv.output}/k8s.ts`);
  }
}


module.exports = new Command();
