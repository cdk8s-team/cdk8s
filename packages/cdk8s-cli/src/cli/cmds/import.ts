import * as yargs from 'yargs';
import { readConfigSync, ImportSpec } from '../../config';
import { importDispatch } from '../../import/dispatch';

const config = readConfigSync();

const DEFAULT_OUTDIR = 'imports';
const LANGUAGES = [ 'typescript', 'python', 'java' ];

class Command implements yargs.CommandModule {
  public readonly command = 'import [SPEC]';
  public readonly describe = 'Imports API objects to your app by generating constructs.';
  public readonly aliases = [ 'gen', 'import', 'generate' ];

  public readonly builder = (args: yargs.Argv) => args
    .positional('SPEC', { default: config.imports, desc: 'import spec with the syntax [NAME:=]SPEC where NAME is an optional module name and supported SPEC are: k8s, crd.yaml, https://domain/crd.yaml).', array: true })
    .example('cdk8s import k8s', 'Imports Kubernetes API objects to imports/k8s.ts')
    .example('cdk8s import k8s@1.13.0', 'Imports a specific version of the Kubernetes API')
    .example('cdk8s import jenkins.io_jenkins_crd.yaml', 'Imports constructs for the Jenkins custom resource definition from a file')
    .example('cdk8s import mattermost:=mattermost_crd.yaml', 'Imports constructs for the mattermost cluster custom resource definition using a custom module name')

    .option('output', { default: DEFAULT_OUTDIR, type: 'string', desc: 'Output directory', alias: 'o' })
    .option('include', { type: 'array', desc: 'Types to select instead of the default which is latest stable version (only for "k8s")' })
    .option('exclude', { type: 'array', desc: 'Do not import types that match these regular expressions. They will be represented as the "any" type (only for "k8s")' })
    .option('language', { default: config.language, demand: true, type: 'string', desc: 'Output programming language', alias: 'l', choices: LANGUAGES });

  public async handler(argv: any) {
    let specs = Array.isArray(argv.spec) ? argv.spec : [ argv.spec ];

    specs = specs.filter((spec: string) => spec != null).map((spec: string) => parseImports(spec));

    await importDispatch(specs, argv, {
      outdir: argv.output,
      targetLanguage: argv.language,
    });
  }
}

function parseImports(spec: string): ImportSpec {
  const splitImport = spec.split(':=');

  // crd.yaml
  // url.com/crd.yaml
  if (splitImport.length === 1) {
    return {
      source: spec,
    }
  }

  // crd=crd.yaml
  // crd=url.com/crd.yaml
  if (splitImport.length === 2) {
    return {
      moduleNamePrefix: splitImport[0],
      source: splitImport[1],
    }
  }

  throw new Error('Unable to parse import specification. Syntax is [NAME:=]SPEC');
}

module.exports = new Command();
