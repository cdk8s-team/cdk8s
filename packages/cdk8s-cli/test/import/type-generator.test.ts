import { TypeGenerator } from "../../lib/import/type-generator";
import { promises as fs } from 'fs';
import * as path from 'path';
import { CodeMaker } from "codemaker";
import { JSONSchema4 } from "json-schema";
import { mkdtemp } from "../../lib/util";
import { srcmak } from "jsii-srcmak";

jest.setTimeout(3 * 60_000); // 1min

describe('unions', () => {

  which('include primitives', {
    oneOf: [
      { type: "string" },
      { type: "number" }
    ]
  });

});


describe('structs', () => {

  which('has primitive types and collections of primitive types', {
    type: 'object',
    properties: {
      stringValue: { type: 'string' },
      boolValue: { type: 'boolean' },
      numberValue: { type: 'number' },
      integerValue: {type: 'integer' },
      arrayOfString: {
        type: 'array',
        items: { type: 'string' }
      }
    }
  });

  which('has a field that references another struct (with required fields)', {
    type: 'object',
    properties: {
      other: {
        $ref: "#/definitions/Other",
      }
    }
  }, {
    definitions: {
      Other: {
        type: 'object',
        properties: {
          stringValue: { type: 'string' }
        },
        required: [ 'stringValue' ]
      }
    }
  });

  which('references itself', {
    type: 'object',
    properties: {
      entrypoint: {
        $ref: '#/definitions/MyType'
      }
    }
  }, {
    definitions: {
      MyType: {
        type: 'object',
        properties: {
          self: { $ref: '#/definitions/MyType' }
        }
      }
    }
  });

  which('array of structs is considered optional', {
    type: 'object',
    properties: {
      shouldBeRequired: { $ref: '#/definitions/ItemType' },
      mapShouldBeOptional: {
        type: 'object',
        additionalProperties: { $ref: '#/definitions/ItemType' }
      },
      arrayShouldBeOptional: {
        type: 'array',
        items: {
          $ref: '#/definitions/ItemType'
        }
      }
    }
  }, {
    definitions: {
      ItemType: {
        type: 'object',
        required: [ 'requiredField' ],
        properties: {
          requiredField: { type: 'string' }
        },
      }
    }
  });

  which('includes required fields', {
    required: [
      "minReadySeconds",
      "revisionHistoryLimit"
    ],
    type: "object",
    properties: {
      minReadySeconds: {
        description: "Minimum number of seconds for which a newly created pod should be ready without any of its container crashing, for it to be considered available. Defaults to 0 (pod will be considered available as soon as it is ready)",
        format: "int32",
        type: "integer"
      },
      paused: {
        description: "Indicates that the deployment is paused.",
        type: "boolean"
      },
      progressDeadlineSeconds: {
        description: "The maximum time in seconds for a deployment to make progress before it is considered to be failed. The deployment controller will continue to process failed deployments and a condition with a ProgressDeadlineExceeded reason will be surfaced in the deployment status. Note that progress will not be estimated during the time a deployment is paused. Defaults to 600s.",
        format: "int32",
        type: "integer"
      },
      replicas: {
        description: "Number of desired pods. This is a pointer to distinguish between explicit zero and not specified. Defaults to 1.",
        format: "int32",
        type: "integer"
      },
      revisionHistoryLimit: {
        description: "The number of old ReplicaSets to retain to allow rollback. This is a pointer to distinguish between explicit zero and not specified. Defaults to 10.",
        format: "int32",
        type: "integer"
      },
    },
  });

});


describe('documentation', () => {

  which('does not render if not defined', {
    type: 'object',
    properties: {
      field: {
        type: 'boolean'
      }
    }
  });

  which('renders based on description', {
    type: 'object',
    properties: {
      field: {
        description: 'hello, description',
        type: 'string'
      }
    }
  });

  which('"*/" is is escaped', {
    type: 'object',
    properties: {
      field: {
        description: 'hello */world',
        type: 'string'
      }
    }
  });

});

describe('enums', () => {

  which('renders a typescript enum', {
    type: 'object',
    required: [ 'firstEnum' ],
    properties: {
      firstEnum: {
        description: 'description of first enum',
        type: 'string',
        enum: [ 'value1', 'value2', 'value-of-three', 'valueOfFour' ]
      },
      child: {
        type: 'object',
        properties: {
          secondEnum: {
            description: 'description of second enum',
            type: 'string',
            enum: [ 'hey', 'enum values can be crazy', 'yes>>123' ]
          }
        }
      }
    }
  });

});

function which(name: string, schema: JSONSchema4, definitions?: JSONSchema4) {
  test(name, async () => {
    const gen = new TypeGenerator(definitions);
    gen.emitType('TestType', schema, 'fqn.of.TestType');

    await mkdtemp(async workdir => {
      expect(await generate(workdir, gen)).toMatchSnapshot();
    });
  });
}

async function generate(workdir: string, gen: TypeGenerator) {
  const code = new CodeMaker();

  const entrypoint = 'index.ts';

  code.openFile(entrypoint);
  gen.generate(code);
  code.closeFile(entrypoint)

  await code.save(workdir);

  const source = await fs.readFile(path.join(workdir, entrypoint), 'utf-8');
  const deps = [ 'constructs', 'cdk8s', '@types/node' ].map(d => path.dirname(require.resolve(`${d}/package.json`)));

  // check that the output compiles & is jsii-compatible
  await srcmak(workdir, { deps });

  return source;
}
