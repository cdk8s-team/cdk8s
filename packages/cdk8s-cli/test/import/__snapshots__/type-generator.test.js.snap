// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentation "*/" is is escaped 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export interface TestType {
    /**
     * hello _/world
     *
     * @schema fqn.of.TestType#field
     */
    readonly field?: string;

}

"
`;

exports[`documentation does not render if not defined 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export interface TestType {
    /**
     * @schema fqn.of.TestType#field
     */
    readonly field?: boolean;

}

"
`;

exports[`documentation renders based on description 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export interface TestType {
    /**
     * hello, description
     *
     * @schema fqn.of.TestType#field
     */
    readonly field?: string;

}

"
`;

exports[`enums renders a typescript enum 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export interface TestType {
    /**
     * description of first enum
     *
     * @schema fqn.of.TestType#firstEnum
     */
    readonly firstEnum: FqnOfTestTypeFirstEnum;

    /**
     * @schema fqn.of.TestType#child
     */
    readonly child?: FqnOfTestTypeChild;

}

/**
 * description of first enum
 *
 * @schema FqnOfTestTypeFirstEnum
 */
export enum FqnOfTestTypeFirstEnum {
    /** value1 */
    VALUE1 = \\"value1\\",
    /** value2 */
    VALUE2 = \\"value2\\",
    /** value-of-three */
    VALUE_OF_THREE = \\"value-of-three\\",
    /** valueOfFour */
    VALUE_OF_FOUR = \\"valueOfFour\\",
}

/**
 * @schema FqnOfTestTypeChild
 */
export interface FqnOfTestTypeChild {
    /**
     * description of second enum
     *
     * @schema FqnOfTestTypeChild#secondEnum
     */
    readonly secondEnum?: FqnOfTestTypeChildSecondEnum;

}

/**
 * description of second enum
 *
 * @schema FqnOfTestTypeChildSecondEnum
 */
export enum FqnOfTestTypeChildSecondEnum {
    /** hey */
    HEY = \\"hey\\",
    /** enum values can be crazy */
    ENUM_VALUES_CAN_BE_CRAZY = \\"enum values can be crazy\\",
    /** yes>>123 */
    YES_123 = \\"yes>>123\\",
}

"
`;

exports[`structs array of structs is considered optional 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export interface TestType {
    /**
     * @schema fqn.of.TestType#shouldBeRequired
     */
    readonly shouldBeRequired?: ItemType;

    /**
     * @schema fqn.of.TestType#mapShouldBeOptional
     */
    readonly mapShouldBeOptional?: { [key: string]: ItemType };

    /**
     * @schema fqn.of.TestType#arrayShouldBeOptional
     */
    readonly arrayShouldBeOptional?: ItemType[];

}

/**
 * @schema ItemType
 */
export interface ItemType {
    /**
     * @schema ItemType#requiredField
     */
    readonly requiredField: string;

}

"
`;

exports[`structs has a field that references another struct (with required fields) 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export interface TestType {
    /**
     * @schema fqn.of.TestType#other
     */
    readonly other?: Other;

}

/**
 * @schema Other
 */
export interface Other {
    /**
     * @schema Other#stringValue
     */
    readonly stringValue: string;

}

"
`;

exports[`structs has primitive types and collections of primitive types 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export interface TestType {
    /**
     * @schema fqn.of.TestType#stringValue
     */
    readonly stringValue?: string;

    /**
     * @schema fqn.of.TestType#boolValue
     */
    readonly boolValue?: boolean;

    /**
     * @schema fqn.of.TestType#numberValue
     */
    readonly numberValue?: number;

    /**
     * @schema fqn.of.TestType#integerValue
     */
    readonly integerValue?: number;

    /**
     * @schema fqn.of.TestType#arrayOfString
     */
    readonly arrayOfString?: string[];

}

"
`;

exports[`structs includes required fields 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export interface TestType {
    /**
     * Minimum number of seconds for which a newly created pod should be ready without any of its container crashing, for it to be considered available. Defaults to 0 (pod will be considered available as soon as it is ready)
     *
     * @default 0 (pod will be considered available as soon as it is ready)
     * @schema fqn.of.TestType#minReadySeconds
     */
    readonly minReadySeconds: number;

    /**
     * Indicates that the deployment is paused.
     *
     * @schema fqn.of.TestType#paused
     */
    readonly paused?: boolean;

    /**
     * The maximum time in seconds for a deployment to make progress before it is considered to be failed. The deployment controller will continue to process failed deployments and a condition with a ProgressDeadlineExceeded reason will be surfaced in the deployment status. Note that progress will not be estimated during the time a deployment is paused. Defaults to 600s.
     *
     * @default 600s.
     * @schema fqn.of.TestType#progressDeadlineSeconds
     */
    readonly progressDeadlineSeconds?: number;

    /**
     * Number of desired pods. This is a pointer to distinguish between explicit zero and not specified. Defaults to 1.
     *
     * @default 1.
     * @schema fqn.of.TestType#replicas
     */
    readonly replicas?: number;

    /**
     * The number of old ReplicaSets to retain to allow rollback. This is a pointer to distinguish between explicit zero and not specified. Defaults to 10.
     *
     * @default 10.
     * @schema fqn.of.TestType#revisionHistoryLimit
     */
    readonly revisionHistoryLimit: number;

}

"
`;

exports[`structs references itself 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export interface TestType {
    /**
     * @schema fqn.of.TestType#entrypoint
     */
    readonly entrypoint?: MyType;

}

/**
 * @schema MyType
 */
export interface MyType {
    /**
     * @schema MyType#self
     */
    readonly self?: MyType;

}

"
`;

exports[`unions include primitives 1`] = `
"/**
 * @schema fqn.of.TestType
 */
export class TestType {
    public static fromString(value: string): TestType {
        return new TestType(value);
    }
    public static fromNumber(value: number): TestType {
        return new TestType(value);
    }
    private constructor(value: any) {
        Object.defineProperty(this, 'resolve', { value: () => value });
    }
}

"
`;
